plugins/ca/root-ca/src/main/java/org/apache/cloudstack/ca/provider/RootCACustomTrustManager.java

```diff
@@ -79,57 +79,58 @@ public void checkClientTrusted(final X509Certificate[] certificates, final Strin
                 LOG.error(errorMsg);
-                throw new CertificateException(errorMsg);                }
-        }
+                throw new CertificateException(errorMsg);
+            }
+
+            // Validity check
+            if (!allowExpiredCertificate) {
+                try {
+                    primaryClientCertificate.checkValidity();
+                } catch (final CertificateExpiredException | CertificateNotYetValidException e) {
+                    final String errorMsg = String.format("Client certificate has expired with serial=%x, subject=%s from address=%s",
+                            primaryClientCertificate.getSerialNumber(), primaryClientCertificate.getSubjectDN(), clientAddress);
+                    LOG.error(errorMsg);
+                    throw new CertificateException(errorMsg);                }
+            }
 
-        // Ownership check
-        boolean certMatchesOwnership = false;
-        if (primaryClientCertificate.getSubjectAlternativeNames() != null) {
-            for (final List<?> list : primaryClientCertificate.getSubjectAlternativeNames()) {
-                if (list != null && list.size() == 2 && list.get(1) instanceof String) {
-                    final String alternativeName = (String) list.get(1);
-                    if (clientAddress.equals(alternativeName)) {
-                        certMatchesOwnership = true;
+            // Ownership check
+            boolean certMatchesOwnership = false;
+            if (primaryClientCertificate.getSubjectAlternativeNames() != null) {
+                for (final List<?> list : primaryClientCertificate.getSubjectAlternativeNames()) {
+                    if (list != null && list.size() == 2 && list.get(1) instanceof String) {
+                        final String alternativeName = (String) list.get(1);
+                        if (clientAddress.equals(alternativeName)) {
+                            certMatchesOwnership = true;
+                        }
                     }
                 }
             }
+            if (!certMatchesOwnership) {
+                final String errorMsg = "Certificate ownership verification failed for client: " + clientAddress;
+                LOG.error(errorMsg);
+                throw new CertificateException(errorMsg);
+            }
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Client/agent connection from ip=" + clientAddress + " has been validated and trusted.");
+            }
         }
-        if (!certMatchesOwnership) {
-            final String errorMsg = "Certificate ownership verification failed for client: " + clientAddress;
-            LOG.error(errorMsg);
-            throw new CertificateException(errorMsg);
-        }
-        if (activeCertMap != null && !Strings.isNullOrEmpty(clientAddress)) {
+        if (primaryClientCertificate != null && activeCertMap != null && !Strings.isNullOrEmpty(clientAddress)) {
```

Comment:
Have you checked the case when auth strictness is disabled and no client certificate (`primaryClientCertificate` is null) ? the in memory map `activeCertMap` stills holds the old client certificate if one exists already in the map.

Reply:
@sureshanaparti the agent handler code on the mgmt server already takes care of removing a cert from the memory map if it existed when the agent disconnects (which would have happened if this code ran at an agent's reconnect).  Also for what it's worth, this part of the functionality is unchanged from the current version of the code.